internal/compiler-bridge/src/main/scala/xsbt/JarUtils.scala

```diff
@@ -0,0 +1,92 @@
+package xsbt
+
+import java.io.File
+import java.util.zip.ZipFile
+
+/**
+ * This is a utility class that provides a set of functions that
+ * are used to implement straight to jar compilation.
+ *
+ * [[sbt.internal.inc.JarUtils]] is an object that has similar purpose and
+ * duplicates some of the code, as it is difficult to share it.
+ */
+final class JarUtils(outputDirs: Iterable[File]) {
+  type ClassInJar = String
+  type RelClass = String
+
+  /**
+   * Creates an identifier for a class located inside a jar.
+   * Mimics the behavior of [[sbt.internal.inc.JarUtils.ClassInJar]].
+   */
+  def ClassInJar(jar: File, cls: RelClass): ClassInJar = {
+    val relClass = if (File.separatorChar == '/') cls else cls.replace('/', File.separatorChar)
+    s"$jar!$relClass"
+  }
+
+  /** Creates an identifier for a class located inside the current output jar. */
+  def ClassInJar(cls: RelClass): ClassInJar = {
+    ClassInJar(outputJar.get, cls)
+  }
+
+  /**
+   * Lists regular files (not directories) inside the given jar.
+   *
+   * @param jar the file to list jars from
+   * @return list of paths to files in jar
+   */
+  def listFiles(jar: File): Set[RelClass] = {
+    import scala.collection.JavaConverters._
+    // ZipFile is slightly slower than IndexBasedZipFsOps but it is quite difficult to use reuse
+    // IndexBasedZipFsOps in compiler bridge.
+    val zip = new ZipFile(jar)
+    try {
+      zip.entries().asScala.filterNot(_.isDirectory).map(_.getName).toSet
+    } finally {
+      zip.close()
+    }
+  }
+
+  /**
+   * The jar file that is used as output for classes. If the output is
+   * not set to a single .jar file, value of this field is [[None]].
+   */
+  val outputJar: Option[File] = {
+    outputDirs match {
+      case Seq(file) if file.getName.endsWith(".jar") => Some(file)
+      case _                                          => None
+    }
+  }
+
+  /**
+   * Informs if the Straight to Jar compilation feature is enabled,
+   * i.e. if the output is set to a jar file.
+   */
+  val isCompilingToJar: Boolean = outputJar.isDefined
+
+  /**
+   * Class that holds cached list of paths located within previous jar for quick lookup.
+   * @see sbt.internal.inc.JarUtils#withPreviousJar for details on what previous jar is
+   */
+  class PrevJarCache(prevJar: Option[File]) extends scala.collection.generic.Clearable {
+    private var cache: Set[ClassInJar] = _
+
+    def contains(classInJar: ClassInJar): Boolean = {
+      if (cache == null) {
+        cache = loadEntriesFromPrevJar()
+      }
+      cache.contains(classInJar)
+    }
+
+    def clear(): Unit = cache = null
+
+    private def loadEntriesFromPrevJar(): Set[ClassInJar] = {
```

Comment:
Now that we have an entry point in the analysis callback, I suggest we use it to remove all this class from the bridge sources. The entry point in analysis callback could look like a replacement for all the heavy lifting you're doing in `loadEntriesFromPrevJar` and `listFiles`. We gain several things from doing this, but I find it important that we don't specify how the references to classes in a jar should look like in the bridge sources (see the implementation of `ClassInJar`). Instead, we should do that decision in the implementation of the analysis callback. Another benefit is the opportunity of using the more efficient `IndexBasedZipFsOps` instead of `ZipFile`. 
  
In fact, I would implement this in the following way:
  
1. Add a `classNamesInJar` method in the callback and remove the one we just added.
2. In the implementation of analysis callback, do. the following:
  2.1. If this is the first compilation, load the entries.
  2.2. If there has been at least one previous compilation, get the entries from a cached store. (We supposedly already know what classes are inside the jar since we used them to decide whether we need to keep compiling incrementally or not)

Reply:
> Add a classNamesInJar method in the callback and remove the one we just added.

Ok, I am starting to see the solution for those problems. Correct me if I am wrong, but as I see it:
We can add `classNamesInJar` method to AnalysisCallback that would return, say, fqns of clases that are currently in jar. It would be a very nice api if in the middle of compilation it would return classes in prev jar and hence be useful for `findAssociatedFile` and in the end (after jvm phase) would return casses in output jar, which would allow us to use it in `Analyzer` (although I am not sure how to implement that correctly, maybe we need 2 methods). This would still require change in the api of callback, like adding a new method, say `generatedLocalClassInJar` or something similar. We can also do th esame for `generatedNonLocalClass` in `API`. 

But going back to `Dependency`, what should we return from `findAssociatedFile` that will be passed to `processExternalDependency` and then to `binaryDependency` and then to the callback.
Currently we return `ClassInJar` wrapped in `PlainFile` which gets us to `binaryDependency` in callback. `classToSource` map holds entries like `ClassInJar (wrapped as File) -> source`, so map lookup works and we correctly record it as internal dependency.
I guess we could also add a variant of `binaryDependency` method to callback to handle this, but what to return from `findAssociatedFile`? `Option[(Either[FqnOfClassInJar, AbstractFile], Boolean)]?

> If there has been at least one previous compilation, get the entries from a cached store
AnalysisCallback is created anew for each scalac run, how should we keep that state?
I have some idea, not sure if this is feasible/safe:
- We have an object, before zinc run it has to be initialized with desired output.
- We will keep fqn cache in that object
- We need to figure out when in the compilation we are: in dependency phase or in analyzer. I think we can get that from recording calls to `dependencyPhaseCompleted` and `get`. This will allow us to know if we should read the jar or use the cache. We might even be able to get rid of the need of looking into prev jar or knowing its path at all. The contents of prev jar will be equal to the contents of output jar from previous run + whatever we are adding after javac which we can also record at the time we are adding it to the output.
- Probably we also need to notify this object when full zinc run is done to clear it

Reply:
I'll answer to this in depth tomorrow, but what I meant is not what you're quite proposing.

Reply:
(Talking about the specifics of the API, not the general idea, which you got right)